master/internal/webhooks/shipper.go

```diff
@@ -0,0 +1,231 @@
+	// System dependencies.
+	log *log.Entry
+
+	// Internal state.
+	wake   chan<- struct{}
+	wg     sync.WaitGroup
+	cancel context.CancelFunc
+}
+
+func newShipper() *shipper {
+	ctx, cancel := context.WithCancel(context.Background()) // Shipper-lifetime scoped context.
+
+	wake := make(chan struct{}, 1)
+	wake <- struct{}{} // Always attempt to process existing events.
+	s := &shipper{
+		log:    log.WithField("component", "webhook-sender"),
+		wake:   wake,
+		cancel: cancel,
+	}
+
+	for i := 0; i < maxWorkers; i++ {
+		s.log.Debugf("creating webhook worker: %d", i)
+		w := newWorker(i)
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+			w.work(ctx, wake)
+		}()
+	}
+
+	return s
+}
+
+// Wake attempts to wake the sender.
+func (s *shipper) Wake() {
+	select {
+	case s.wake <- struct{}{}:
+	default:
+		// If the channel is full, we will forgo sending. We will already wake, and that will
+		// be after we persisted the event that caused this wake.
+		// The only critical correctness condition to be aware of is: if you wake, you must consume
+		// all events. If you stop short, they may be delay until the next wake.
+	}
+}
+
+func (s *shipper) Close() {
+	s.cancel()
+	s.wg.Wait()
+}
+
+func newWorker(id int) *worker {
+	return &worker{
+		log: log.WithFields(log.Fields{"component": "webhook-shipper-worker", "id": id}),
+		cl:  cleanhttp.DefaultClient(),
+	}
+}
+
+type worker struct {
+	// System dependencies.
+	log *logrus.Entry
+	cl  *http.Client
+}
+
+func (w *worker) work(ctx context.Context, wake <-chan struct{}) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			w.log.Errorf("uncaught error, webhook worker crashed: %v", rec)
+		}
+	}()
+
+	for {
+		select {
+		case <-wake:
+			if err := w.ship(ctx); err != nil {
+				w.log.WithError(err).Error("failed to ship batch")
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *worker) ship(ctx context.Context) error {
+loop:
+	for {
+		switch n, err := w.shipBatch(ctx); {
+		case err != nil:
+			return err
+		case n <= 0:
+			return nil
+		default:
+			// Continue until events are exhausted.
+			continue loop
+		}
+	}
+}
+
+func (w *worker) shipBatch(ctx context.Context) (int, error) {
+	b, err := dequeueEvents(ctx, maxEventBatchSize)
+	if err != nil {
+		return 0, fmt.Errorf("getting events: %w", err)
+	}
+	defer func() {
+		if err := b.close(); err != nil {
+			w.log.WithError(err).Warn("failed to finalize batch")
+		}
+	}()
+
+	var wg sync.WaitGroup
+	for _, e := range b.events {
+		wg.Add(1)
+		go func(e Event) {
+			defer wg.Done()
+			if err := backoff.Retry(
+				func() error { return w.deliver(ctx, e) },
+				w.backoff(),
+			); err != nil {
+				w.log.WithError(err).Error("failed to deliver webhook")
+			}
+		}(e)
+	}
+	wg.Wait()
+	if err := ctx.Err(); err != nil {
+		return 0, err
+	}
+
+	if err := b.consume(); err != nil {
+		return 0, fmt.Errorf("consuming batch: %w", err)
+	}
+	return len(b.events), nil
+}
+
+func (w *worker) backoff() backoff.BackOff {
```

Comment:
doesn't have to be a worker method
```suggestion
func backoff() backoff.BackOff {
```

Reply:
updated